cpp/src/io/json/json_tree.cu

```diff
@@ -0,0 +1,661 @@
+  //   a. sort by level
+  //   b. get gather map of sorted indices
+  //   c. translate parent_node_ids to sorted indices
+  rmm::device_uvector<size_type> scatter_indices(num_nodes, stream);
+  thrust::sequence(rmm::exec_policy(stream), scatter_indices.begin(), scatter_indices.end());
+#ifdef NJP_DEBUG_PRINT
+  printf("\n");
+  print_vec(cudf::detail::make_std_vector_async(scatter_indices, stream), "gpu.node_id");
+  print_vec(cudf::detail::make_std_vector_async(d_tree.parent_node_ids, stream),
+            "gpu.parent_node_ids");
+  print_vec(cudf::detail::make_std_vector_async(node_type, stream), "gpu.node_type");
+  print_vec(cudf::detail::make_std_vector_async(d_tree.node_levels, stream), "gpu.node_levels");
+#endif
+  rmm::device_uvector<NodeIndexT> parent_node_ids(d_tree.parent_node_ids, stream);  // make a copy
+  auto out_pid =
+    thrust::make_zip_iterator(scatter_indices.data(), parent_node_ids.data(), node_type.data());
+  // Uses cub radix sort.
+  thrust::stable_sort_by_key(rmm::exec_policy(stream),
+                             d_tree.node_levels.data(),
+                             d_tree.node_levels.data() + num_nodes,
+                             out_pid);
+  auto gather_indices = cudf::detail::scatter_to_gather(
+    scatter_indices.begin(), scatter_indices.end(), num_nodes, stream);
+
+  rmm::device_uvector<NodeIndexT> parent_indices(num_nodes, stream);
+  *thrust::device_pointer_cast(parent_indices.data()) = -1;
+  thrust::gather(rmm::exec_policy(stream),
+                 parent_node_ids.begin() + 1,  // first node's parent is -1
+                 parent_node_ids.end(),
+                 gather_indices.begin(),
+                 parent_indices.begin() + 1);
+#ifdef NJP_DEBUG_PRINT
+  printf("\n");
+  print_vec(cudf::detail::make_std_vector_async(scatter_indices, stream), "gpu.node_id");
+  print_vec(cudf::detail::make_std_vector_async(parent_node_ids, stream), "gpu.parent_node_ids");
+  print_vec(cudf::detail::make_std_vector_async(node_type, stream), "gpu.node_type");
+  print_vec(cudf::detail::make_std_vector_async(d_tree.node_levels, stream), "gpu.node_levels");
+  print_vec(cudf::detail::make_std_vector_async(gather_indices, stream), "new_home");
+  print_vec(cudf::detail::make_std_vector_async(parent_indices, stream), "parent_indices");
+  print_vec(cudf::detail::make_std_vector_async(parent_node_ids, stream),
+            "parent_node_ids (restored)");
+#endif
+  // 3. Find level boundaries.
+  hostdevice_vector<size_type> level_boundaries(num_nodes + 1, stream);
+  auto level_end = thrust::copy_if(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<size_type>(1),
+    thrust::make_counting_iterator<size_type>(num_nodes + 1),
+    level_boundaries.d_begin(),
+    [num_nodes, node_levels = d_tree.node_levels.begin()] __device__(auto index) {
+      return index == 0 || index == num_nodes || node_levels[index] != node_levels[index - 1];
```

Comment:
Is the comparison against `0` needed? We begin the `copy_if` from index `1` onwards, right?
```suggestion
      return index == num_nodes || node_levels[index] != node_levels[index - 1];
```

Reply:
0 is not required. `0` was added with meaning `boundaries` in mind.
will remove this.